enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+
+`Cloud-controller` will be provisioned in the `openshift-cloud-controller-manager` namespace.
+
+Preparation steps, common for all providers:
+
+- Add support for recognition of the `external` option in [library-go](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) which will disable config parsing and add support for `external` field in the `Infrastructure` resource.
+- Port the templates of `cloud-controller` from upstream repositories into proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository. Those templates will belong to the newly created `openshift-cloud-controller-manager` namespace and be managed with an operator.
+- Operator will be deployed in the `openshift-cloud-controller-manager-operator` namespace.
+
+Each provider implementation will need to do the following minimal set of actions in order to make out-of-tree implemetation work:
+
+- Use `external` flag in the `kube-controller-manager` pod instead of the previously used provider. Remove the `cloud-config` option from the template.
+- Use `external` flag in the `kubelet` pod. Remove the `cloud-config` option from the template.
+- Provision a `cloud-controller-manager` `DaemonSet` in the `openshift-cloud-controller-manager` namespace.
+
+#### API resource changes
+
+In order to achieve desired cluster configuration with `external` cloud components, we need to add a new field into `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11). With the addition, a `status.platformStatus.external` field will determine the architectural change in the cluster configuration. Other components will be able to consume the configuration same way as before, only with a minor change of the resource processing.
+
+```go
+type PlatformStatus struct {
+  ...
+	// External is a value used in infrastucture automaton to determine component-specific
+	// behavior for the PlatformType value. Components depending on this variable will
+	// ignore cloud specific detailes associated with the given platform type,
+	// and will concider the value of the infrastructure to be "external" - essentially
+	// disabling any further evaluation.
+	//
+	// This will allow to delegate task of processing platform specific bits
+	// to another component.
+	// +optional
+  External bool `json:"external"`
+  ...
+```
+
+To ensure a smooth transition from the `in-tree` selection, `library-go` [implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) will include an additional argument for the `cloudProviderObserver` structure, which will allow to preserve initial configuration strategy, until the new `observeExternal` flag will be set to `false`:
+
+```go
+type cloudProviderObserver struct {
```

Comment:
I'm not sure why `observeExternal ` this is needed, we got all we need by watching at `status.platformStatus.external` to decide what to set for `cloudProviderConfigPath`

Reply:
Is this the mechanism proposed to enable backward compatible changes during the development transition?

Reply:
Yes, this will be the first step to transition to our operator management. We will use this, so we won't diverge from the existing approach of watching and copying the `cloud-config` configMap into the namespace, where it should be managed. This is currently used by kube-api-server and kube-controller-manager operators. Later this will be substituted by a CR as you suggested, similar to the `KubeControllerManager` one.

Reply:
So it was decided this implementation will be for development/testing only. The final migration will be done by a set of changes on all the repos, reliant on the selected provider.